### PR TITLE
Icon in "Open with" menus

### PR DESCRIPTION
**Pull request by jeremypw**
Icons are added to the "Open with" submenus in both the pathbar and view.

The simplest (smallest diff) method is used for each.  In the pathbar a grid is added to a GtkMenuItem.  In the view, a GLib.MenuItem is used, linked to the appropriate pre-existing action.

If necessary, GLibMenuItems and actions can be used in the pathbar too.

The unnecessary use of null instead of an empty list in a related function is removed.

### CODE REVIEW

File: src/View/AbstractDirectoryView.vala [hunk 1]
```diff
@@ -2103,26 +2102,27 @@ namespace FM {
...
-                                count++;
-                            }
-
-                            last_label = label.dup ();
-                            last_exec = exec.dup ();
+                        /* Ensure no duplicate items */
+                        unowned string label = app_info.get_display_name ();
+                        unowned string exec = app_info.get_executable ().split (" ")[0];
+                        if (label != last_label || exec != last_exec) {
+                             var menu_item = new GLib.MenuItem (app_info.get_name (), "selection.open_with_app::" + index.to_string ());
```
Comment (tintou) [inline src/View/AbstractDirectoryView.vala]: Just a nice thing that is also possible :
```suggestion
                             var menu_item = new GLib.MenuItem (app_info.get_name (), GLib.Action.print_detailed_name ("selection.open_with_app", new GLib.Variant.int32 (index)));
```

**Comment by jeremypw** [comment]
Thanks for the suggestion!

File: src/View/AbstractDirectoryView.vala [hunk 2]
```diff
@@ -2103,26 +2102,27 @@ namespace FM {
...
+                            menu_item.set_icon (app_info.get_icon ());
+                            apps_section.append_item (menu_item);
+                            count++;
                         }
+
+                        last_label = label.dup ();
```
Comment (tintou) [inline src/View/AbstractDirectoryView.vala]: Vala will automatically duplicate the string if it is `unowned string` to `string`
```suggestion
                        last_label = label;
```

File: src/View/AbstractDirectoryView.vala [hunk 3]
```diff
@@ -2103,26 +2102,27 @@ namespace FM {
...
+                        last_exec = exec.dup ();
```
Comment (tintou) [inline src/View/AbstractDirectoryView.vala]:
```suggestion
                        last_exec = exec;
```

File: src/Utils/MimeActions.vala [hunk 4]
```diff
@@ -171,23 +167,19 @@ public class Marlin.MimeActions {
             GLib.List<AppInfo> one_result = get_applications_for_file (file);
             one_result.sort (application_compare_by_id);
 
-            if (result != null) {
+            if (result.length () > 0) {
```
Comment (tintou) [inline src/Utils/MimeActions.vala]: length iterates the whole list when you can do:
```suggestion
            if (result.data != null) {
```

File: src/Utils/MimeActions.vala [hunk 5]
```diff
@@ -171,23 +167,19 @@ public class Marlin.MimeActions {
...
                 result = intersect_application_lists (result, one_result);
             } else {
                 result = (owned) one_result;
             }
 
-            if (result == null) {
-                return null;
+            if (result.length () == 0) {
```
Comment (tintou) [inline src/Utils/MimeActions.vala]: length iterates the whole list when you can do:
```suggestion
            if (result.data == null) {
```

File: src/View/AbstractDirectoryView.vala [hunk 6]
```diff
@@ -2103,26 +2102,27 @@ namespace FM {
...
-                if (open_with_apps != null) {
+                if (open_with_apps.length () > 0) {
                     var apps_section = new GLib.Menu ();
                     int index = -1;
                     int count = 0;
                     string last_label = "";
                     string last_exec = "";
 
-                    foreach (var app in open_with_apps) {
+                    foreach (var app_info in open_with_apps) {
```
Comment (tintou) [inline src/View/AbstractDirectoryView.vala]: GLib.List allows unowned iteration, better use it to not add reference count without any reason ;)
```suggestion
                    foreach (unowned GLib.AppInfo app_info in open_with_apps) {
```

File: src/View/AbstractDirectoryView.vala [hunk 7]
```diff
@@ -2103,26 +2102,27 @@ namespace FM {
 
                 filter_this_app_from_open_with_apps ();
 
-                if (open_with_apps != null) {
+                if (open_with_apps.length () > 0) {
```
Comment (tintou) [inline src/View/AbstractDirectoryView.vala]:
```suggestion
                if (open_with_apps.data != null) {
```

File: src/View/Widgets/BreadcrumbsEntry.vala [hunk 8]
```diff
@@ -464,22 +464,19 @@ namespace Marlin.View.Chrome {
             foreach (AppInfo app_info in app_info_list) {
                 if (app_info != null && app_info.get_executable () != Environment.get_application_name ()) {
                     at_least_one = true;
-                    var menu_item = new Gtk.ImageMenuItem.with_label (app_info.get_name ());
+                     var item_grid = new Gtk.Grid ();
+                    item_grid.add (new Gtk.Image.from_gicon (app_info.get_icon (), Gtk.IconSize.MENU));
```
Comment (tintou) [inline src/View/Widgets/BreadcrumbsEntry.vala]: I have the feeling that you'll need to also specify `pixel_size` to not allow the image to be bigger than Gtk.IconSize.MENU (happened in the past with Firefox and LibreOffice, to verify)

File: src/View/AbstractDirectoryView.vala (hunk 2 above)
Comment (jeremypw) [reply src/View/AbstractDirectoryView.vala]: OK, thanks. Maybe an earlier Vala version didn't do that?

File: src/View/AbstractDirectoryView.vala (hunk 3 above)
Comment (jeremypw) [reply src/View/AbstractDirectoryView.vala]: I'll commit all these changes in one PR

File: src/Utils/MimeActions.vala (hunk 4 above)
Comment (jeremypw) [reply src/Utils/MimeActions.vala]: In this case the length is only a handful of apps so I wasn't concerned but I'll make these changes.

**Comment by tintou** [comment]
> @tintou: Regarding pixel_size property of Image, do you need to consider the display scale-factor when setting this?

No it will scale it accordingly

**Pull request merged**